src/evo/mnauth.cpp

```diff
@@ -0,0 +1,141 @@
+{
+    if (!fMasternodeMode) {
+        return;
+    }
+
+    if (!masternodeSync.IsBlockchainSynced()) {
+        // we can't really verify MNAUTH messages when we don't have the latest MN list
+        return;
+    }
+
+    if (strCommand == NetMsgType::MNAUTH) {
+        CMNAuth mnauth;
+        vRecv >> mnauth;
+
+        {
+            LOCK(pnode->cs_mnauth);
+            // only one MNAUTH allowed
+            if (!pnode->verifiedProRegTx.IsNull()) {
+                LOCK(cs_main);
+                Misbehaving(pnode->id, 100);
+                return;
+            }
+        }
+
+        if (mnauth.proRegTxHash.IsNull() || !mnauth.sig.IsValid()) {
+            LOCK(cs_main);
+            Misbehaving(pnode->id, 100);
+            return;
+        }
+
+        auto mnList = deterministicMNManager->GetListAtChainTip();
+        auto dmn = mnList.GetValidMN(mnauth.proRegTxHash);
+        if (!dmn) {
+            LOCK(cs_main);
+            Misbehaving(pnode->id, 10);
+            // in case he was unlucky and not up to date, let him retry the whole verification process
+            pnode->fDisconnect = true;
+            return;
+        }
+
+        uint256 signHash;
+        {
+            LOCK(pnode->cs_mnauth);
+            // See comment in PushMNAUTH (fInbound is negated here as we're on the other side of the connection)
+            signHash = ::SerializeHash(std::make_tuple(dmn->pdmnState->pubKeyOperator, pnode->sentMNAuthChallenge, !pnode->fInbound));
+        }
+
+        if (!mnauth.sig.VerifyInsecure(dmn->pdmnState->pubKeyOperator, signHash)) {
+            LOCK(cs_main);
+            Misbehaving(pnode->id, 10);
+            // in case he was unlucky and not up to date, let him retry the whole verification process
+            pnode->fDisconnect = true;
+            return;
+        }
+
+        connman.ForEachNode([&](CNode* pnode2) {
+            if (pnode2->verifiedProRegTx == mnauth.proRegTxHash) {
+                LogPrint("net", "CMNAuth::ProcessMessage -- Masternode %s has already verified as peer %d, dropping old connection. peer=%d\n",
+                        mnauth.proRegTxHash.ToString(), pnode2->id, pnode->id);
+                pnode2->fDisconnect = true;
+            }
+        });
+
+        {
+            LOCK(pnode->cs_mnauth);
+            pnode->verifiedProRegTx = mnauth.proRegTxHash;
+            pnode->verifiedPubKey = dmn->pdmnState->pubKeyOperator.GetHash();
+        }
+
+        LogPrint("net", "CMNAuth::%s -- Valid MNAUTH for %s, peer=%d\n", __func__, mnauth.proRegTxHash.ToString(), pnode->id);
+    }
+}
+
+void CMNAuth::NotifyMasternodeListChanged(const CDeterministicMNList& newList)
+{
+    std::unordered_set<uint256> pubKeys;
+    g_connman->ForEachNode([&](const CNode* pnode) {
+        LOCK(pnode->cs_mnauth);
+        if (!pnode->verifiedProRegTx.IsNull()) {
+            pubKeys.emplace(pnode->verifiedProRegTx);
```

Comment:
```suggestion
            pubKeys.emplace(pnode->verifiedPubKey);
```

Reply:
Whoops :see_no_evil: Thanks for catching this one.